File: clients/android/app/src/main/java/com/onlineclipboard/app/core/LocalStore.kt
package com.onlineclipboard.app.core

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

class LocalStore(context: Context) : SQLiteOpenHelper(context, "local.db", null, 1) {
    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL("CREATE TABLE meta(k TEXT PRIMARY KEY, v BLOB)")
        db.execSQL(
            """CREATE TABLE clips(
                id TEXT PRIMARY KEY, status TEXT, version INTEGER, created_seq TEXT,
                created_at TEXT, expires_at TEXT, source_device_id TEXT,
                vault_id TEXT, nonce TEXT, ciphertext TEXT, plaintext TEXT)"""
        )
        db.execSQL("CREATE TABLE outbox(id TEXT PRIMARY KEY, envelope_json TEXT, created_at TEXT)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {}

    fun getString(key: String): String? {
        readableDatabase.rawQuery("SELECT v FROM meta WHERE k=?", arrayOf(key)).use { c ->
            if (!c.moveToFirst()) return null
            return c.getBlob(0)?.toString(Charsets.UTF_8)
        }
    }

    fun getBytes(key: String): ByteArray? {
        readableDatabase.rawQuery("SELECT v FROM meta WHERE k=?", arrayOf(key)).use { c ->
            if (!c.moveToFirst()) return null
            return c.getBlob(0)
        }
    }

    fun set(key: String, value: String) = set(key, value.toByteArray(Charsets.UTF_8))

    fun set(key: String, value: ByteArray) {
        writableDatabase.execSQL(
            "INSERT INTO meta(k,v) VALUES(?,?) ON CONFLICT(k) DO UPDATE SET v=excluded.v",
            arrayOf(key, value)
        )
    }

    fun upsertClip(item: HistoryItem, vaultId: String, nonce: String, ciphertext: String) {
        writableDatabase.execSQL(
            """INSERT INTO clips(id,status,version,created_seq,created_at,expires_at,source_device_id,vault_id,nonce,ciphertext,plaintext)
               VALUES(?,?,?,?,?,?,?,?,?,?,?)
               ON CONFLICT(id) DO UPDATE SET status=excluded.status, version=excluded.version,
                 created_seq=excluded.created_seq, expires_at=excluded.expires_at,
                 plaintext=excluded.plaintext, nonce=excluded.nonce, ciphertext=excluded.ciphertext""",
            arrayOf(
                item.id, item.status, item.version, item.createdSeq, item.createdAt, item.expiresAt ?: "",
                item.sourceDeviceId, vaultId, nonce, ciphertext, LocalProtect.wrap(item.text.toByteArray(Charsets.UTF_8))
            )
        )
    }

    fun deleteClip(id: String) {
        writableDatabase.execSQL("DELETE FROM clips WHERE id=?", arrayOf(id))
    }

    fun loadClips(status: String): List<HistoryItem> {
        val list = mutableListOf<HistoryItem>()
        readableDatabase.rawQuery(
            "SELECT id,status,version,created_seq,created_at,expires_at,source_device_id,plaintext FROM clips WHERE status=? ORDER BY created_seq DESC",
            arrayOf(status)
        ).use { c ->
            while (c.moveToNext()) {
                val text = readPlain(c, 7)
                list.add(
                    HistoryItem(
                        id = c.getString(0),
                        status = c.getString(1),
                        version = c.getInt(2),
                        createdSeq = c.getString(3),
                        createdAt = c.getString(4),
                        expiresAt = c.getString(5).ifEmpty { null },
                        sourceDeviceId = c.getString(6),
                        text = text,
                        preview = previewOf(text)
                    )
                )
            }
        }
        return list
    }

    fun addOutbox(id: String, json: String) {
        writableDatabase.execSQL(
            "INSERT OR REPLACE INTO outbox(id,envelope_json,created_at) VALUES(?,?,?)",
            arrayOf(id, json, System.currentTimeMillis().toString())
        )
    }

    fun outbox(): List<Pair<String, String>> {
        val list = mutableListOf<Pair<String, String>>()
        readableDatabase.rawQuery("SELECT id, envelope_json FROM outbox ORDER BY created_at", null).use { c ->
            while (c.moveToNext()) list.add(c.getString(0) to c.getString(1))
        }
        return list
    }

    fun removeOutbox(id: String) {
        writableDatabase.execSQL("DELETE FROM outbox WHERE id=?", arrayOf(id))
    }

    fun outboxCount(): Int {
        readableDatabase.rawQuery("SELECT COUNT(*) FROM outbox", null).use { c ->
            return if (c.moveToFirst()) c.getInt(0) else 0
        }
    }

    private fun readPlain(c: android.database.Cursor, i: Int): String {
        if (c.isNull(i)) return ""
        return try {
            val blob = c.getBlob(i)
            String(LocalProtect.unwrap(blob), Charsets.UTF_8)
        } catch (_: Exception) {
            c.getString(i) ?: ""
        }
    }

    companion object {
        fun previewOf(text: String): String {
            val one = text.replace("\r", " ").replace("\n", " ")
            return if (one.length <= 80) one else one.take(80) + "…"
        }
    }
}

data class HistoryItem(
    val id: String,
    var status: String,
    var version: Int,
    var createdSeq: String,
    val createdAt: String,
    var expiresAt: String?,
    val sourceDeviceId: String,
    var text: String,
    var preview: String
) {
    override fun toString(): String = preview
}

object LocalProtect {
    private const val ALIAS = "onlineclipboard-local"

    private fun key(): SecretKey {
        val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        (ks.getKey(ALIAS, null) as? SecretKey)?.let { return it }
        val gen = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
        gen.init(
            KeyGenParameterSpec.Builder(
                ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build()
        )
        return gen.generateKey()
    }

    fun wrap(data: ByteArray): ByteArray {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.ENCRYPT_MODE, key())
        return c.iv + c.doFinal(data)
    }

    fun unwrap(data: ByteArray): ByteArray {
        require(data.size > 12) { "wrapped local secret too short" }
        val iv = data.copyOfRange(0, 12)
        val ct = data.copyOfRange(12, data.size)
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.DECRYPT_MODE, key(), GCMParameterSpec(128, iv))
        return c.doFinal(ct)
    }
}

File: clients/windows/OnlineClipboard.Windows/Core/LocalStore.cs
using System.IO;
using Microsoft.Data.Sqlite;

namespace OnlineClipboard.Windows.Core;

public sealed class LocalStore : IDisposable
{
    private readonly SqliteConnection _db;

    public LocalStore(string path)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        _db = new SqliteConnection(new SqliteConnectionStringBuilder { DataSource = path }.ToString());
        _db.Open();
        using var cmd = _db.CreateCommand();
        cmd.CommandText = """
            CREATE TABLE IF NOT EXISTS meta(k TEXT PRIMARY KEY, v BLOB);
            CREATE TABLE IF NOT EXISTS clips(
                id TEXT PRIMARY KEY, status TEXT, version INTEGER, created_seq TEXT,
                created_at TEXT, expires_at TEXT, source_device_id TEXT,
                vault_id TEXT, nonce TEXT, ciphertext TEXT, plaintext TEXT);
            CREATE TABLE IF NOT EXISTS outbox(
                id TEXT PRIMARY KEY, envelope_json TEXT, delivery_intent TEXT, created_at TEXT);
            CREATE TABLE IF NOT EXISTS ops(
                id TEXT PRIMARY KEY, clip_id TEXT, kind TEXT, version INTEGER);
            """;
        cmd.ExecuteNonQuery();
    }

    public string? GetString(string key)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "SELECT v FROM meta WHERE k=$k";
        cmd.Parameters.AddWithValue("$k", key);
        return cmd.ExecuteScalar() as string;
    }

    public byte[]? GetBytes(string key)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "SELECT v FROM meta WHERE k=$k";
        cmd.Parameters.AddWithValue("$k", key);
        return cmd.ExecuteScalar() as byte[];
    }

    public void Set(string key, object value)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "INSERT INTO meta(k,v) VALUES($k,$v) ON CONFLICT(k) DO UPDATE SET v=excluded.v";
        cmd.Parameters.AddWithValue("$k", key);
        cmd.Parameters.AddWithValue("$v", value);
        cmd.ExecuteNonQuery();
    }

    public void UpsertClip(HistoryItem item, string vaultId, string nonce, string ciphertext)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = """
            INSERT INTO clips(id,status,version,created_seq,created_at,expires_at,source_device_id,vault_id,nonce,ciphertext,plaintext)
            VALUES($id,$st,$ver,$seq,$ca,$ex,$src,$vault,$n,$c,$p)
            ON CONFLICT(id) DO UPDATE SET status=excluded.status, version=excluded.version,
                created_seq=excluded.created_seq, expires_at=excluded.expires_at, plaintext=excluded.plaintext,
                nonce=excluded.nonce, ciphertext=excluded.ciphertext
            """;
        cmd.Parameters.AddWithValue("$id", item.Id);
        cmd.Parameters.AddWithValue("$st", item.Status);
        cmd.Parameters.AddWithValue("$ver", item.Version);
        cmd.Parameters.AddWithValue("$seq", item.CreatedSeq);
        cmd.Parameters.AddWithValue("$ca", item.CreatedAt.ToString("O"));
        cmd.Parameters.AddWithValue("$ex", item.ExpiresAt?.ToString("O") ?? (object)DBNull.Value);
        cmd.Parameters.AddWithValue("$src", item.SourceDeviceId);
        cmd.Parameters.AddWithValue("$vault", vaultId);
        cmd.Parameters.AddWithValue("$n", nonce);
        cmd.Parameters.AddWithValue("$c", ciphertext);
        cmd.Parameters.AddWithValue("$p", CryptoV1.ProtectLocal(System.Text.Encoding.UTF8.GetBytes(item.Text)));
        cmd.ExecuteNonQuery();
    }

    public void DeleteClip(string id)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "DELETE FROM clips WHERE id=$id";
        cmd.Parameters.AddWithValue("$id", id);
        cmd.ExecuteNonQuery();
    }

    public List<HistoryItem> LoadClips(string status)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "SELECT id,status,version,created_seq,created_at,expires_at,source_device_id,plaintext FROM clips WHERE status=$s ORDER BY created_seq DESC";
        cmd.Parameters.AddWithValue("$s", status);
        using var r = cmd.ExecuteReader();
        var list = new List<HistoryItem>();
        while (r.Read())
        {
            var text = ReadPlain(r, 7);
            list.Add(new HistoryItem
            {
                Id = r.GetString(0),
                Status = r.GetString(1),
                Version = r.GetInt32(2),
                CreatedSeq = r.GetString(3),
                CreatedAt = DateTimeOffset.Parse(r.GetString(4)),
                ExpiresAt = r.IsDBNull(5) ? null : DateTimeOffset.Parse(r.GetString(5)),
                SourceDeviceId = r.GetString(6),
                Text = text,
                Preview = PreviewOf(text)
            });
        }
        return list;
    }

    public HistoryItem? GetClip(string id)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "SELECT id,status,version,created_seq,created_at,expires_at,source_device_id,plaintext FROM clips WHERE id=$id";
        cmd.Parameters.AddWithValue("$id", id);
        using var r = cmd.ExecuteReader();
        if (!r.Read()) return null;
        var text = ReadPlain(r, 7);
        return new HistoryItem
        {
            Id = r.GetString(0), Status = r.GetString(1), Version = r.GetInt32(2), CreatedSeq = r.GetString(3),
            CreatedAt = DateTimeOffset.Parse(r.GetString(4)),
            ExpiresAt = r.IsDBNull(5) ? null : DateTimeOffset.Parse(r.GetString(5)),
            SourceDeviceId = r.GetString(6), Text = text, Preview = PreviewOf(text)
        };
    }

    public void AddOutbox(string id, string json)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "INSERT OR REPLACE INTO outbox(id,envelope_json,created_at) VALUES($id,$j,$t)";
        cmd.Parameters.AddWithValue("$id", id);
        cmd.Parameters.AddWithValue("$j", json);
        cmd.Parameters.AddWithValue("$t", DateTimeOffset.UtcNow.ToString("O"));
        cmd.ExecuteNonQuery();
    }

    public List<(string Id, string Json)> Outbox()
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "SELECT id, envelope_json FROM outbox ORDER BY created_at";
        using var r = cmd.ExecuteReader();
        var list = new List<(string, string)>();
        while (r.Read()) list.Add((r.GetString(0), r.GetString(1)));
        return list;
    }

    public void RemoveOutbox(string id)
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "DELETE FROM outbox WHERE id=$id";
        cmd.Parameters.AddWithValue("$id", id);
        cmd.ExecuteNonQuery();
    }

    public int OutboxCount()
    {
        using var cmd = _db.CreateCommand();
        cmd.CommandText = "SELECT COUNT(*) FROM outbox";
        return Convert.ToInt32(cmd.ExecuteScalar());
    }

    private static string ReadPlain(SqliteDataReader r, int i)
    {
        if (r.IsDBNull(i)) return "";
        var value = r.GetValue(i);
        if (value is byte[] blob)
        {
            try { return System.Text.Encoding.UTF8.GetString(CryptoV1.UnprotectLocal(blob)); }
            catch { return ""; }
        }
        return value.ToString() ?? "";
    }

    public static string PreviewOf(string text)
    {
        var one = text.Replace("\r", " ").Replace("\n", " ");
        return one.Length <= 80 ? one : one[..80] + "…";
    }

    public void Dispose() => _db.Dispose();
}
